server/routes/api/gmail-auth.js

```diff
@@ -0,0 +1,156 @@
+const { google } = require("googleapis");
+const mailComposer = require("nodemailer/lib/mail-composer"); // Helps formatting of emails in base64
+const User = require("../../models/user.js");
+const { client_secret, client_id } = require("../../gmail-secret.json");
+
+/*
+ * Scopes dictate what we are allowed to do on behalf of the user and what the user is asked to approve
+ * If modifying these scopes, existing tokens will need to be deleted
+ */
+const SCOPES = [
+  "https://www.googleapis.com/auth/gmail.readonly",
+  "https://www.googleapis.com/auth/gmail.compose",
+  "https://www.googleapis.com/auth/gmail.send",
+];
+
+// Url user redirected to after Google authorization
+// If modified it also needs to be changed at https://console.developers.google.com/apis/credentials?project=mail-sender-1
+const SUCCESS_REDIRECT_URL = "http://lvh.me:3001/api/gmail-auth/success";
+
+/*************
+ * TODO: Get UID of current logged in user, likely from req.header
+ **************/
+const userId = "5e827c5ed999b93a267ef847";
+
+/*
+ * Generate url to redirect user to for authorization on Google and get new token.
+ */
+function getNewTokenUrl() {
+  const oAuth2Client = new google.auth.OAuth2(client_id, client_secret, SUCCESS_REDIRECT_URL);
+  const authUrl = oAuth2Client.generateAuthUrl({
+    access_type: "offline",
+    scope: SCOPES,
+    // ONLY NEEDED IN DEV to ensure refresh token is provided every time token is requested
+    ...(!process.env.NODE_ENV && { prompt: "consent" }),
+  });
+  return authUrl;
+}
+
+/**
+ * Check if gmail authorization token exists for the user, if not, get a new one
+ */
+exports.checkForToken = async (req, res, next) => {
+  try {
+    const loggedInUser = await User.findById(userId);
+
+    if (loggedInUser.gmail_token) {
+      console.log("Token exists");
+
+      /***********
+       * TODO: Redirect back to user home page
+       ***********/
+      res.redirect("/"); // OR next(req);
+    } else {
+      console.log("no token");
+      res.redirect(getNewTokenUrl());
+    }
+  } catch (error) {
+    console.log(error);
+  }
+};
+
+/*
+ * When user has authorized Mail Sender with Gmail, user is redirected here
+ * Saves new token to database under that user
+ */
+exports.processToken = async (req, res) => {
+  try {
+    const oAuth2Client = new google.auth.OAuth2(client_id, client_secret, SUCCESS_REDIRECT_URL);
+    const { tokens } = await oAuth2Client.getToken(req.query.code);
+
+    // Store the token to db under user
+    const user = await User.findById(userId);
+    user.gmail_token = tokens.refresh_token;
+    await user.save().then((res, err) => {
+      if (err) {
+        console.log("Error writing token to users collection:", err);
+        res.status(500).redirect("/");
+      }
+    });
+
+    /***********
+     * TODO: Redirect back to user home page
+     ***********/
+    res.redirect("/");
```

Comment:
Since we are using a single page application, we should actually be moving some of this redirecting logic into the React app - this is something we can discuss in standup

For now this is good for this PR, but something we'll have to change when you do the front-end of this

Reply:
I realised this when planning the front-end last night :) doh!